doc/docs/master/how-tos/developer-workflow/build-pipelines.md

```diff
@@ -0,0 +1,118 @@
+
+1. `/pfs/build` - any artifacts resulting from the build process.
+
+1. `/pfs/<input(s)>` - any inputs specified in the pipeline specification.
+
+## Builders
+The builder is responsible for determining what Docker image is used for the base image and what steps to run during the pipeline build and deployment. If a builder has a standard implementation (currently `python` and `go`), then only the `transform.build.language` needs to be set, as it has impelmentations for `build.sh` and `run.sh`. The order of preference is:
+
+1. `transform.build.language`
+1. `transform.build.image`
+1. `transform.image`
+
+If `transform.buil.language` is not specified, then the build pipeline is non-standard and uses the `transform.build.image` container (and subsequently the `transform.build` if `transform.build.image` is not provided). In a non-standard implementation, the convention is to provide `build.sh` and `run.sh` scripts to fulfill the build pipeline requirements. 
+
+Similarly, a build pipeline by default will run `sh /pfs/build/run.sh` by default. If a `transform.cmd` is specified, it will take precedence over `run.sh`.
+
+### Python Builder
+
+The Python builder requires a file structure similar to the following:
+
+```tree
+./map
+├── source
+│   ├── requirements.txt
+│   ├── ...
+│   └── main.py
+└── build-pipeline.json
+```
+There must exist a `main.py` which acts as the entrypoint for the pipeline. Optionally, a `requirements.txt` can be used to specify pip packages that will be installed during the build process. Other supporting files in the directory will also be copied and available in the pipeline if they are not excluded by the `.pachignore`.
+
+The `build.sh` and `run.sh` files are optional, as the Python Builder already contains these scripts. However, if the base image is modified, they must be provided.
+
+### Go Builder
+
+The Go Builder follows the same format as the [Python Builder](#python-builder). There must be a main source file in the source root that imports and invokes the intended code.
+
+### Creating a Builder
+
+For languages other than Python and Go, or customizations to the official builders, users can author their own builders. Builders are somewhat similar to buildpacks in design, and follow a convention over configuration approach. A builder needs 3 things:
+
+- A Dockerfile to bake the image specified in the build pipeline spec.
+- A `build.sh` in the image workdir, which acts as the entry-point for the build pipeline.
+- A `run.sh`, injected into `/pfs/out` via `build.sh`. This will act as the entry-point for the executing pipeline. By convention, `run.sh` should take an arbitrary number of arguments and forward them to whatever executes the actual user code.
+
+The build file structure would have the following format: 
+```tree
+./project
+├── source
+│   ├── requirements.txt
+│   ├── main.py
+│   ├── build.sh
+│   └── run.sh
+└── build-pipeline.json
+```
```

Comment:
This isn't right, see the official builders in `etc/pipeline-build` for guiding examples. In particular, there's no pipeline spec, `main.py` or `requirements.txt` in either of them.

Reply:
I think I worded it poorly here. This is supposed to be the local project structure where they are developing, i.e. the pipeline they are developing, not the create their own builder. So they would have to have a pipeline spec.

Reply:
I'll look at it and try to clarify.